iwf/workflow_state.py

```diff
@@ -0,0 +1,108 @@
+                    Note that any write API will be recorded to server after the whole waitUntil API response is accepted
+            communication: the API right now only for publishing value to internalChannel
+                        Note that any write API will be recorded to server after the whole waitUntil API response is accepted.
+
+        Returns: the requested command
+        """
+        pass
+
+    @abstractmethod
+    def execute(
+        self,
+        ctx: WorkflowContext,
+        input: T,
+        command_results: CommandResults,
+        persistence: Persistence,
+        communication: Communication,
+    ) -> StateDecision:
+        """
+        Execute is the method to execute and decide what to do next. Invoke after commands from WaitUntil are completed, or there is WaitUntil is not implemented for the state.
+
+          Args:
+              ctx: the context info of this API invocation, like workflow start time, workflowId, etc
+              input:  the state input
+              command_results: the results of the command that executed by WaitUntil
+              persistence:  the API for
+                    1) data attributes: defined by ObjectWorkflow interface
+                    2) search attributes: defined by ObjectWorkflow interface
+                    3) stateExecutionLocals: for passing data within the state execution
+                    4) recordEvent: for storing some tracking info(e.g. RPC call input/output) when executing the API.
+                    Note that any write API will be recorded to server after the whole waitUntil API response is accepted
+              communication: the API right now only for publishing value to internalChannel.
+                            Note that any write API will be recorded to server after the whole execute API response is accepted.
+
+          Returns: the decision of what to do next(e.g. transition to next states or closing workflow)
+
+        """
+        raise NotImplementedError
+
+    def get_state_options(self) -> WorkflowStateOptions:
+        """GetStateOptions can just return nil to use the default Options
+        StateOptions is optional configuration to adjust the state behaviors. Default values:
+             StateId:  name of the implementation class
+             waitUntilApiFailurePolicy: FAIL_WORKFLOW_ON_FAILURE
+             PersistenceLoadingPolicy for dataAttributes/searchAttributes: LOAD_ALL_WITHOUT_LOCKING
+             waitUntil/execute API:
+                timeout: 30s
+                retryPolicy:
+                    InitialIntervalSeconds: 1
+                    MaxInternalSeconds:100
+                    MaximumAttempts: 0
+                    BackoffCoefficient: 2
+        Returns: WorkflowStateOptions
+        """
+        pass
+
+
+def get_state_id(state: WorkflowState) -> str:
+    options = state.get_state_options()
+    if options is None or options.state_id is None:
+        return state.__class__.__name__
```

Comment:
should this be a fully qualified name? https://gist.github.com/clbarnes/edd28ea32010eb159b34b075687bb49e